### DISCUSSION TITLE
My First Analyzer... but in another projet don't found Collection.Immutable  8.0.0.0

### DISCUSSION THREAD

**Discussion by Nidkfr** [dotnet/roslyn#71989] [Q&A]
Hi,

   I created an analyzer in a net8 project latest version of VS (17.8.6). All works, all unit tests work. But when I put the package in another project I have an error CS8032 for System.Collection.Immutable 8.0.0.0. I saw that in the VS folder it uses 7.0.0.0
I can't find a solution to make this work. Do you have any solutions?

**Comment by jasonmalinowski** [comment]
Although it's possible to bundle 8.0 you might run into some issues that way; can you just use 7.0? Visual Studio 17.9 Previews do now ship 8.0.

**Reply by Nidkfr** [reply]
I actually had this idea with 17.9 after posting this message, now I no longer have the error with this Dll but I have a TypeInitializerEXception. But I can't get the innerexception.
I try to pass to 4.9.final but same error. I had not constructor juste static initialize

**Reply by jasonmalinowski** [reply]
> but I have a TypeInitializerEXception

Can you share that?

**Reply by Nidkfr** [reply]
what do you need?

**Reply by jasonmalinowski** [reply]
The exception message or something? Is this from the command line build or something else?

**Reply by Nidkfr** [reply]
I use french version
1>C:\_rd\workspaces\[Dev Annexe]\Akanea\Akanea.Ressources\Akanea.Ressources\Ressources\Stores\JsonRessourcesStore.cs(51,26,51,41): warning CS1572: Le commentaire XML a une balise param pour 'pServiceLocator', alors qu'il n'existe aucun paramètre de ce nom
1>CSC : warning AD0001: L'analyseur 'Akanea.Analyzer.CallingMethodNeedOnMultiLineAnalyzer' a levé une exception de type 'System.TypeInitializationException' avec le message 'Une exception a été levée par l'initialiseur de type pour 'Akanea.Analyzer.CallingMethodNeedOnMultiLineAnalyzer'.'.
1>Akanea.Ressources -> C:\_rd\workspaces\[Dev Annexe]\Akanea\Akanea.Ressources\Akanea.Ressources\bin\Debug\net8.0\Akanea.Ressources.dll
1>Génération du projet "Akanea.Ressources.csproj" terminée.

**Reply by jasonmalinowski** [reply]
Hmm, tagging @jaredpar as general feedback that if we have an analyzer exception, we really should print inner exceptions too.

**Reply by jasonmalinowski** [reply]
Why do you have a static constructor though?

**Reply by CyrusNajmabadi** [reply]
note: i've been able to find the exception stacks in binlogs.  I do agree that it would be nice to see from the command line itself.

**Reply by jasonmalinowski** [reply]
@CyrusNajmabadi Yeah that was my thought; no reason to force users to play "hunt the exception" if we can just be more helpful.

**Reply by Nidkfr** [reply]
it's my source code : 

```
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

using System.Collections.Immutable;

namespace Akanea.Analyzer
{
    
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class CallingMethodNeedOnMultiLineAnalyzer : DiagnosticAnalyzer
    {
        #region Fields

        private static readonly DiagnosticDescriptor _rule =
           new(
               DIAGNOSTIC_ID,
               DIAGNOSTIC_TITLE,
               DIAGNOSTIC_MESSAGE_FORMAT,
               DIAGNOSTIC_CATEGORY,
               DiagnosticSeverity.Warning,
               isEnabledByDefault: true,
               description: DIAGNOSTIC_DESCRIPTION);

        public const string DIAGNOSTIC_CATEGORY = "Akanea";
        public const string DIAGNOSTIC_DESCRIPTION = "Improve reading code.";
        public const string DIAGNOSTIC_ID = "AK0001";
        public const string DIAGNOSTIC_MESSAGE_FORMAT = "the calling methode {0} need split parameters for reading";
        public const string DIAGNOSTIC_TITLE = "When you call methode split parameter if is more than 2 parameters";

        #endregion Fields

        #region Properties

       
        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
        { get { return [_rule]; } }

        #endregion Properties

        #region Private Methods

        private void _analyzeInvocationExpression(SyntaxNodeAnalysisContext pContext)
        {
            if (pContext.Node is InvocationExpressionSyntax expression)
            {
                if (pContext.SemanticModel.GetSymbolInfo(expression).Symbol is IMethodSymbol symbol)
                {
                    int countParameter = symbol.Parameters.Length;
                    int countReturn = 0;
                    foreach (SyntaxToken token in expression.ArgumentList.ChildTokens().Where(pToken => pToken.IsKind(SyntaxKind.CommaToken)))
                    {
                        if (token.GetAllTrivia().Any(pTrivia => pTrivia.IsKind(SyntaxKind.EndOfLineTrivia)))
                            countReturn++;
                    }

                    if (countParameter > 2 && countParameter - 1 > countReturn)
                    {
                        pContext.ReportDiagnostic(Diagnostic.Create(_rule, expression.GetLocation(), symbol.Name));
                    }
                }
            }
        }

        #endregion Private Methods

        #region Public Methods

   
        public override void Initialize(AnalysisContext pContext)
        {
            pContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
            pContext.EnableConcurrentExecution();
            pContext.RegisterSyntaxNodeAction(_analyzeInvocationExpression, SyntaxKind.InvocationExpression);
        }

        #endregion Public Methods
    }
}
```

**Reply by jasonmalinowski** [reply]
So are you still referencing System.Collections.Immutable 8.0 or something lower?

**Reply by Nidkfr** [reply]
my projet reference:
 ```
<PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.8.0" />
 <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.8.0" />
 <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.Common" Version="4.8.0" />
 <PackageReference Include="System.Collections.Immutable" Version="8.0.0" />
```

**Reply by jasonmalinowski** [reply]
Just try downgrading the System.Collections.Immutable back to 7.0, you might still be picking up other things that aren't expecting the newer dependency. You also shouldn't have a reference to Workspaces.Common on the analyzer DLL itself.

**Reply by Nidkfr** [reply]
for the workspace lib i make in same dll analyzer and codefix. I change all (comment all codefix code) change pakage
i have same erreur AD001 , typeInitializerException
```
<PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.8.0" />
<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.8.0" />
<PackageReference Include="System.Collections.Immutable" Version="7.0.0" />
```

**Reply by jasonmalinowski** [reply]
Are you building with an older .NET SDK?

**Reply by Nidkfr** [reply]
normally I took the latest one: dotnet-sdk-8.0.101-win-x64. but I couldn't see if it's really this one that is used

**Reply by Nidkfr** [reply]
I took my project out of my solution in .net8 and moved it into the project template provided by visual, updated the packages and positioned myself in netStandard 2.0. Now everything works. I will stay in this configuration. I modified the old project to no longer have static initialization, and it shows me the real exception. It's still not possible to load the System.Immutable 8.0.0.0 library on vs17.8.3
On the other hand, on 17.9preview, it tells me that I cannot load System.runtime 8.0.0.0

`CSC : warning AD0001: L'analyseur 'Akanea.Analyzer.CallingMethodNeedOnMultiLineAnalyzer' a levé une exception de type 'System.IO.FileNotFoundException' avec le message 'Impossible de charger le fichier ou l'assembly 'System.Runtime, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' ou une de ses dépendances. Le fichier spécifié est introuvable.'.`

**Reply by jasonmalinowski** [reply]
Oh yeah, analyzers have to be netstandard. And don't reference System.Collections.Immutable if you don't have to, and also reference the lowest possible Microsoft.CodeAnalysis you can. If you're referencing 4.8 that means you can't be loaded in any Visual Studio lower than 17.8. That may not be a problem for you if this is a private analyzer for your team, but if you plan on more broadly distributing, you may want to target a lower version.
